src/lib/core/CHIPTLVText.cpp

```diff
@@ -0,0 +1,231 @@
+        case kTLVType_Array:
+        case kTLVType_List:
+        {
+            TLVType containerType;
+            const char *openingBrace, *closingBrace;
+
+            if (reader.GetType() == kTLVType_Structure) {
+                openingBrace = "{";
+                closingBrace = "}";
+            }
+            else if (reader.GetType() == kTLVType_Array) {
+                openingBrace = "[";
+                closingBrace = "]";
+            }
+            else {
+                openingBrace = "[[";
+                closingBrace = "]]";
+            }
+
+            if (IsContextTag(reader.GetTag()) || IsProfileTag(reader.GetTag())) {
+                printerFunc("\n");
+                PrintTabs(printerFunc, tabLevel);
+            }
+
+            printerFunc("%s\n", openingBrace);
+
+            err = reader.EnterContainer(containerType);
+            SuccessOrExit(err);
+
+            isFirst = true;
+            while ((err = reader.Next()) == CHIP_NO_ERROR) {
+                err = Print(reader, printerFunc, tabLevel + 1, isFirst);
+                SuccessOrExit(err);
+                isFirst = false;
+            }
+
+            err = reader.ExitContainer(containerType);
+            SuccessOrExit(err);
+
+            printerFunc("\n");
+            PrintTabs(printerFunc, tabLevel);
+            printerFunc("%s", closingBrace);
+            
+            if (tabLevel == 0) {
+                printerFunc("\n");
+            }
+
+            break;
+        }
+
+        case kTLVType_Boolean:
+        {
+            bool v;
+
+            err = reader.Get(v);
+            SuccessOrExit(err);
+
+            printerFunc("%s", (v) ? "true" : "false");
+            break;
+        }
+
+        case kTLVType_SignedInteger:
+        {
+            int64_t v;
+
+            err = reader.Get(v);
+            SuccessOrExit(err);
+
+            printerFunc("%" PRId64 "", v);
+            break;
+        }
+
+        case kTLVType_FloatingPointNumber:
+        {
+            double f;
+
+            err = reader.Get(f);
+            SuccessOrExit(err);
+
+            printerFunc("%.4f", f);
+            break;
+        }
+
+        case kTLVType_UnsignedInteger:
+        {
+            uint64_t v;
+
+            err = reader.Get(v);
+            SuccessOrExit(err);
+
+            printerFunc("%" PRIu64 "", v);
+            break;
+        }
+
+        case kTLVType_ByteString:
+        {
+            uint32_t sz = reader.GetLength();
+            uint8_t buf[sz];
+
+            err = reader.GetBytes(buf, sz);
+            SuccessOrExit(err);
+
+            for (uint32_t i = 0; i < sz; i++) {
+                if (i != 0) {
+                    printerFunc(" ");
+                }
+
+                printerFunc("%" PRIxLEAST8 "", buf[i]);
+            }
+
+            break;
+        }
+
+        case kTLVType_UTF8String:
+        {
+            uint32_t sz = reader.GetLength();
+            char buf[sz + 1];
```

Comment:
Again, this is not safe for untrusted TLV.

Reply:
Have switched over to using a Span instead that points directly into the TLV buffer.